File: 2016/01/01.py
def parse_file(file):
    with open(file) as f:
        return [
            parse_step(step)     
            for step in f.readline().strip().split(', ')
        ]


def parse_step(step):
    return step[0], int(step[1:])


def turn(direction, turn):
    if turn == 'R':
        if direction == 'N':
            return 'E'
        if direction == 'E':
            return 'S'
        if direction == 'S':
            return 'W'
        if direction == 'W':
            return 'N'

    if turn == 'L':
        if direction == 'N':
            return 'W'
        if direction == 'E':
            return 'N'
        if direction == 'S':
            return 'E'
        if direction == 'W':
            return 'S'


def walk(steps):
    x, y = 0, 0
    seen = set((x, y))
    current_direction = 'N'
    first_seen_twice = None
    for direction, length in steps:
        dx, dy = 0, 0
        current_direction = turn(current_direction, direction)
        if current_direction == 'N':
            dy = 1
        if current_direction == 'S':
            dy = -1
        if current_direction == 'E':
            dx = 1
        if current_direction == 'W':
            dx = -1

        for i in range(length):
            x += dx
            y += dy
            if (x, y) in seen and first_seen_twice is None:
                first_seen_twice = x, y
            seen.add((x, y))
    
    return distance((x, y)), distance(first_seen_twice)


def distance(point):
    x, y = point
    return abs(x) + abs(y)


steps = parse_file('input.txt')
one, two = walk(steps)
print('Part 1:', one)
print('Part 2:', two)

File: 2016/01/01.go
package main

import "fmt"
import "strings"
import "io/ioutil"
import "strconv"

type Step struct {
	direction string
	steps     int
}

type Point struct {
	x, y int
}

func parse(file string) []Step {
	steps := []Step{}
	data, _ := ioutil.ReadFile(file)
	for _, line := range strings.Split(strings.Trim(string(data), "\n"), ", ") {
		if len(line) == 0 {
			continue
		}
		parts := strings.SplitN(line, "", 2)
		stepCount, _ := strconv.Atoi(parts[1])
		steps = append(steps, Step{parts[0], stepCount})
	}
	return steps
}

func turn(current, direction string) string {
	if direction == "R" {
		switch current {
		case "E":
			return "S"
		case "S":
			return "W"
		case "W":
			return "N"
		case "N":
			return "E"
		}
	} else {
		switch current {
		case "E":
			return "N"
		case "S":
			return "E"
		case "W":
			return "S"
		case "N":
			return "W"
		}
	}
	return ""
}

func follow(start Point, steps []Step) Point {
	direction := "E"
	for _, step := range steps {
		direction = turn(direction, step.direction)
		switch direction {
		case "E":
			start.x += step.steps
		case "W":
			start.x -= step.steps
		case "N":
			start.y += step.steps
		case "S":
			start.y -= step.steps
		}
	}
	return start
}

func contains(ps []Point, p Point) bool {
	for _, pp := range ps {
		if pp.x == p.x && pp.y == p.y {
			return true
		}
	}
	return false
}

func firstVisitTwice(start Point, steps []Step) Point {
	visited := []Point{}
	direction := "E"
	for _, step := range steps {
		direction = turn(direction, step.direction)
		for i := 0; i < step.steps; i++ {
			switch direction {
			case "E":
				start.x++
			case "W":
				start.x--
			case "N":
				start.y++
			case "S":
				start.y--
			}
			if contains(visited, start) {
				return start
			}
			visited = append(visited, start)
		}
	}
	return Point{-1, -1}
}

func main() {
	steps := parse("input.txt")
	end := follow(Point{0, 0}, steps)
	fmt.Println("Part 1:", end.x+end.y)

	vis := firstVisitTwice(Point{0, 0}, steps)
	fmt.Println("Part 2:", vis.x+vis.y)
}
